src/quickgui/qgsquickmapcanvasmap.cpp

```diff
@@ -0,0 +1,394 @@
+/***************************************************************************
+  qgsquickmapcanvasmap.cpp
+  --------------------------------------
+  Date                 : 10.12.2014
+  Copyright            : (C) 2014 by Matthias Kuhn
+  Email                : matthias (at) opengis.ch
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QQuickWindow>
+#include <QScreen>
+#include <QSGSimpleTextureNode>
+#include <QtConcurrent>
+
+#include "qgsmaprendererparalleljob.h"
+#include "qgsmessagelog.h"
+#include "qgspallabeling.h"
+#include "qgsproject.h"
+#include "qgsvectorlayer.h"
+#include "qgis.h"
+
+#include "qgsquickmapcanvasmap.h"
+#include "qgsquickmapsettings.h"
+
+
+QgsQuickMapCanvasMap::QgsQuickMapCanvasMap( QQuickItem *parent )
+  : QQuickItem( parent )
+  , mMapSettings( new QgsQuickMapSettings() )
+{
+  connect( this, &QQuickItem::windowChanged, this, &QgsQuickMapCanvasMap::onWindowChanged );
+  connect( &mRefreshTimer, &QTimer::timeout, this, &QgsQuickMapCanvasMap::refreshMap );
+  connect( &mMapUpdateTimer, &QTimer::timeout, this, &QgsQuickMapCanvasMap::renderJobUpdated );
+
+  Q_ASSERT( mMapSettings );
+  connect( mMapSettings, &QgsQuickMapSettings::extentChanged, this, &QgsQuickMapCanvasMap::onExtentChanged );
+  connect( mMapSettings, &QgsQuickMapSettings::layersChanged, this, &QgsQuickMapCanvasMap::onLayersChanged );
+
+  connect( this, &QgsQuickMapCanvasMap::renderStarting, this, &QgsQuickMapCanvasMap::isRenderingChanged );
+  connect( this, &QgsQuickMapCanvasMap::mapCanvasRefreshed, this, &QgsQuickMapCanvasMap::isRenderingChanged );
+
+  mMapUpdateTimer.setSingleShot( false );
+  mMapUpdateTimer.setInterval( 250 );
+  mRefreshTimer.setSingleShot( true );
+  setTransformOrigin( QQuickItem::TopLeft );
+  setFlags( QQuickItem::ItemHasContents );
+}
+
+QgsQuickMapSettings *QgsQuickMapCanvasMap::mapSettings() const
+{
+  Q_ASSERT( mMapSettings );
```

Comment:
I can't understand all these asserts, how will this ever be a `nullptr`?

Reply:
If you do not mind, I would like to keep it here. It is because of all other components that use map settings do have code like 
`if (mMapSettings) {...}`. So during development, I got confused few times why here there is no such code. So i explicitly defining this `Q_ASSERT` in the beginning of each function to state that it is always defined in this context (no need to look around in the code and trying to find out)

Reply:
removed (not needed with unique_ptr)